tests/_utils.py

```diff
@@ -2,41 +2,151 @@
+        expected = np.array(expected, dtype=np.float32)
+    np.testing.assert_array_equal(actual, expected, strict=strict)
+
+
+parametrize = pytest.mark.parametrize
+
+
+def default(value: Any) -> FixtureFunction:
+    """Create a ``fixture`` that returns a default value.
+
+    When using nested ``fixture``'s, this helps simulate "optional" parameters; test cases can explicitly set values for
+    fixtures, but also omit them (which PyTest normally errors over).
+    """
+    return fixture(lambda: value)
+
+
+def param(**kwargs):
+    """``kwargs``-based wrapper around ``parametrize``, for fixtures that are constant in a given test case.
+
+    Analogous to declaring "constant" variables in the fixture DAG. This is useful/common when some ``fixture``'s depend
+    on other ``fixture``'s, and one or more of the dependencies have a fixed value in all "parametrizations" of a given
+    test case (e.g. when other parameters are being "swept" over).
+
+    For example:
+
+    >>> @param(name='Alice', phone='111-111-1111')
+    ... @sweep(dob=['1/1/2000', '1/1/3000'])  # Test multiple DOBs with this user name/phone
+    ... def test_case(user):  # Fixture `user` is constructed from `name`, `phone`, and `dob`
+    ...     ...
+
+    Is equivalent to:
+
+    >>> @parametrize("name,phone", [('Alice', '111-111-1111')])
+    ... @parametrize("dob", [['1/1/2000', '1/1/3000']])  # Test multiple DOBs with this user name/phone
+    ... def test_case(user):  # Fixture `user` is constructed from `name`, `phone`, and `dob`
+    ...     ...
+    """
+
+    def rv(fn):
+        for k, v in reversed(kwargs.items()):
+            fn = parametrize(f"{k}", [v])(fn)
+
+        return fn
+
+    return rv
+
+
+def sweep(**kwargs):
+    """``kwargs``-based wrapper around ``parametrize``, for fixtures that are not constant in a given test case, e.g.:
+
+    >>> @sweep(n=[1, 2, 3])
+    ... def test_case(n):
+    ...     ...
+
+    Is equivalent to:
 
+    >>> @parametrize("n", [1, 2, 3])
+    ... def test_case(n):
+    ...     ...
+    """
+
+    def rv(fn):
+        for k, v in reversed(kwargs.items()):
+            fn = parametrize(f"{k}", v)(fn)
+
+        return fn
+
+    return rv
 
-def pytorch_seq_x_value_gen(obs_range: range, var_range: range) -> spmatrix:
-    """A sparse matrix where the values of each col are the obs_range values.
 
-    Useful for checking the X values are being returned in the correct order.
+def coords_to_float(r: int, c: int) -> float:
+    """Encode a (row,col) position as a float.
+
+    ``r`` is placed to the left of the decimal, while ``c`` is placed to the right with its digits reversed (otherwise
+    ``(10,10)`` and ``(10,1)`` would collide).
     """
-    data = np.vstack([list(obs_range)] * len(var_range)).flatten()
-    rows = np.vstack([list(obs_range)] * len(var_range)).flatten()
-    cols = np.column_stack([list(var_range)] * len(obs_range)).flatten()
+    return float(f"{r}.{str(c)[::-1]}")
```

Comment:
Could we use another method here instead of reversing? Like scaling for example since the former might create some misinterpretations.

Reply:
> one "scaling" option would be r + c * 1e-4

imho this looks cleaner but I will leave it to your judgement.